### PR TITLE
Add remove(Sync)/removeAll(Sync)

### PR DESCRIPTION
**Pull request by kevinkassimo**
- [x] Add `remove/removeSync/removeAll/removeAllSync`
- [x] Add tests

For `remove/removeSync/removeAll/removeAllSync`,
+ Node: `fs.rmdirSync(path)`
+ Go: `os.RemoveAll(path string) error`
+ Python: `os.removedirs(path)`
+ Rust: `std::fs::remove_dir_all<P: AsRef<Path>>(path: P) -> Result<()>`
(similar for async version)

### CODE REVIEW

File: src/fs.rs [hunk 1]
```diff
@@ -49,6 +48,18 @@ pub fn mkdir(path: &Path) -> std::io::Result<()> {
   })
 }
 
+pub fn rmdir(path: &Path) -> std::io::Result<()> {
+  debug!("rmdir {}", path.display());
+  std::fs::remove_dir_all(path).or_else(|err| {
+    // Ignore NotFound error
+    if err.kind() == std::io::ErrorKind::NotFound {
```
Comment (NewLunarFire) [inline src/fs.rs]: I prefer match expressions for those. They are shorter and more idiomatic Rust.

match err.kind() {
    std::io::ErrorKind::NotFound => Ok(),
    _ => Err(err)
}

File: src/handlers.rs [hunk 2]
```diff
@@ -556,6 +567,46 @@ macro_rules! to_seconds {
...
+fn handle_mkdir_sync(
+  _d: *const DenoC,
+  _builder: &mut FlatBufferBuilder,
+  _dirname: &str,
+  _perm: u32,
+) -> HandlerResult {
+  debug!("handle_mkdir_sync {}", _dirname);
+  let deno = from_c(_d);
+  if deno.flags.allow_write {
+    // TODO: apply permission
```
Comment (NewLunarFire) [inline src/handlers.rs]: Look at the code in #567 .

Comment (kevinkassimo) [reply src/handlers.rs]: Thanks for the reference! Have just added a similar way to set permission with `DirBuilderExt`

File: src/fs.rs (hunk 1 above)
Comment (kevinkassimo) [reply src/fs.rs]: Done, thanks!

**Comment by ry** [comment]
Waiting to land #708 before this one.

**Comment by kevinkassimo** [comment]
@ry updated based on #708 changes.

File: src/fs.rs [hunk 3]
```diff
@@ -37,14 +40,37 @@ pub fn make_temp_dir(
...
+}
+
+#[cfg(not(any(unix)))]
+fn set_dir_permission(_builder: &mut DirBuilder, _perm: u32) {
+  // NOOP on windows
+}
+
+pub fn rmdir(path: &Path) -> std::io::Result<()> {
+  debug!("rmdir {}", path.display());
+  remove_dir_all(path).or_else(|err| {
```
[the remark this answers is not in the document]
Comment (kevinkassimo) [reply src/fs.rs]: Done. Added a new optional param `recursive`, default to `false`

File: js/rmdir.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Removes a directory with the specified path.
+ *
+ *     import { rmdir } from "deno";
+ *     await rmdir("dir");
+ */
+export async function rmdir(path: string, recursive = false): Promise<void> {
+  await dispatch.sendAsync(...req(path, recursive));
+}
```
Comment (ry) [inline js/rmdir.ts]: In terms of the public API, I think we should have separate functions for the recursive function. Internally you can use a flag to reduce code.

Please change the API:
- remove() and removeSync()   (currently called rmdir and rmdirSync)
- removeAll() and removeAllSync (same but with recursive flag)

And please this file to js/remove.ts

File: src/handlers.rs [hunk 5]
```diff
@@ -420,17 +421,32 @@ fn handle_make_temp_dir(d: *const DenoC, base: &msg::Base) -> Box<Op> {
...
   }
   // TODO Use tokio_threadpool.
   Box::new(futures::future::result(|| -> OpResult {
     debug!("handle_mkdir {}", path);
-    // TODO(ry) Use mode.
-    deno_fs::mkdir(Path::new(path))?;
+    deno_fs::mkdir(Path::new(path), mode)?;
+    Ok(None)
+  }()))
+}
```
Comment (ry) [inline src/handlers.rs]: Leave out changes to mkdir - do these in a separate commit.

File: src/fs.rs [hunk 6]
```diff
@@ -64,14 +67,40 @@ pub fn make_temp_dir(
...
+  debug!("rmdir {}", path.display());
+  let res = if recursive {
+    remove_dir_all(path)
+  } else {
+    remove_dir(path)
+  };
+  res.or_else(|err| {
+    match err.kind() {
+      // Ignore NotFound error
+      std::io::ErrorKind::NotFound => Ok(()),
```
Comment (ry) [inline src/fs.rs]: Don't ignore NotFound error.

File: src/fs.rs [hunk 7]
```diff
@@ -64,14 +67,40 @@ pub fn make_temp_dir(
...
+  debug!("set dir perm to {}", perm);
+  builder.mode(perm);
+}
+
+#[cfg(not(any(unix)))]
+fn set_dir_permission(_builder: &mut DirBuilder, _perm: u32) {
+  // NOOP on windows
+}
+
+pub fn rmdir(path: &Path, recursive: bool) -> std::io::Result<()> {
```
Comment (ry) [inline src/fs.rs]: Remove this function, everything should be handled in `handle_remove` inside src/handlers.rs

File: src/fs.rs [hunk 8]
```diff
@@ -64,14 +67,40 @@ pub fn make_temp_dir(
...
+#[cfg(any(unix))]
+fn set_dir_permission(builder: &mut DirBuilder, perm: u32) {
+  debug!("set dir perm to {}", perm);
+  builder.mode(perm);
+}
+
+#[cfg(not(any(unix)))]
+fn set_dir_permission(_builder: &mut DirBuilder, _perm: u32) {
+  // NOOP on windows
+}
```
Comment (ry) [inline src/fs.rs]: Leave out permission changes. Let's do that in a separate PR.

File: src/msg.fbs [hunk 9]
```diff
@@ -173,6 +174,11 @@ table Mkdir {
   // mode specified by https://godoc.org/os#FileMode
 }
 
+table Rmdir {
+  path: string;
+  recursive: bool;
+}
```
Comment (ry) [inline src/msg.fbs]: Rename to `Remove`

File: js/rmdir.ts (hunk 4 above)
Comment (kevinkassimo) [reply js/rmdir.ts]: Sure, feels more readable. Although I do feel probably something like `removeDir` would be clearer? (to avoid confusing with removing a file)

Currently at office. Would finish working on this by evening.

Comment (ry) [reply js/rmdir.ts]: In libc "remove" handles both directories and files: [remove(3)](https://linux.die.net/man/3/remove)

Comment (kevinkassimo) [reply js/rmdir.ts]: Got it. Will combine the 2 usages into a single function.

File: src/fs.rs (hunk 8 above)
Comment (kevinkassimo) [reply src/fs.rs]: Restored to original mkdir. Will submit another PR for it.

File: src/fs.rs (hunk 7 above)
Comment (kevinkassimo) [reply src/fs.rs]: Done.

**Comment by kevinkassimo** [comment]
@ry Done. (Hope I did not miss something)
